src/analysis/processing/qgsalgorithmorderbyexpression.cpp

```diff
@@ -0,0 +1,96 @@
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#include "qgsalgorithmorderbyexpression.h"
+#include "qgsfeaturerequest.h"
+
+
+///@cond PRIVATE
+
+QString QgsOrderByExpressionAlgorithm::name() const
+{
+  return QStringLiteral( "orderbyexpression" );
+}
+
+QString QgsOrderByExpressionAlgorithm::displayName() const
+{
+  return QObject::tr( "Order by expression" );
+}
+
+QStringList QgsOrderByExpressionAlgorithm::tags() const
+{
+  return QObject::tr( "orderby,sort,expression,field" ).split( ',' );
+}
+
+QString QgsOrderByExpressionAlgorithm::group() const
+{
+  return QObject::tr( "Vector general" );
+}
+
+void QgsOrderByExpressionAlgorithm::initAlgorithm( const QVariantMap & )
+{
+  addParameter( new QgsProcessingParameterFeatureSource( QStringLiteral( "INPUT" ), QObject::tr( "Input layer" ) ) );
+  addParameter( new QgsProcessingParameterExpression( QStringLiteral( "EXPRESSION" ), QObject::tr( "Expression" ), QVariant(), QStringLiteral( "INPUT" ) ) );
+  addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "ASCENDING" ), QObject::tr( "Ascending" ), true ) );
+  addParameter( new QgsProcessingParameterBoolean( QStringLiteral( "NULLS_FIRST" ), QObject::tr( "Nulls first" ), false ) );
+
+  addParameter( new QgsProcessingParameterFeatureSink( QStringLiteral( "OUTPUT" ), QObject::tr( "Output layer" ) ) );
+}
+
+QString QgsOrderByExpressionAlgorithm::shortHelpString() const
+{
+  return QObject::tr( "This algorithm sorts a vector layer according to an expression." );
+}
+
+QgsOrderByExpressionAlgorithm *QgsOrderByExpressionAlgorithm::createInstance() const
+{
+  return new QgsOrderByExpressionAlgorithm();
+}
+
+QVariantMap QgsOrderByExpressionAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  std::unique_ptr< QgsFeatureSource > source( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !source )
+    return QVariantMap();
+
+  QString expressionString = parameterAsExpression( parameters, QStringLiteral( "EXPRESSION" ), context );
+
+  bool ascending = parameterAsBool( parameters, QStringLiteral( "ASCENDING" ), context );
+  bool nullsFirst = parameterAsBool( parameters, QStringLiteral( "DROP_M_VALUES" ), context );
+
+  QString sinkId;
+  std::unique_ptr< QgsFeatureSink > sink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, sinkId, source->fields(), source->wkbType(), source->sourceCrs() ) );
+  if ( !sink )
+    return QVariantMap();
+
+  QgsFeatureRequest request;
+  request.addOrderBy( expressionString, ascending, nullsFirst );
+
+  QgsFeature inFeature;
+  QgsFeatureIterator features = source->getFeatures( request );
+  while ( features.nextFeature( inFeature ) )
+  {
+    sink->addFeature( inFeature );
```

Comment:
Add feedback for progress add cancellation here... Someone might be sorting a million row table!